File: marketing/web10-social/src/data/access.ts
// ── access.ts — access recovery ─────────────────────────────────────────────
// Executes the confirmatory access verdict from w.verifyAccess() (the node is
// the oracle — see knowledge/knowledge-base/web10-v3/sdk/api.md "Access
// Health"). It never guesses from status codes; it runs the verdict's ordered
// `actions`, honoring a cooldown so a transient false positive can't churn the
// user into a re-auth loop.
//
// Generic oracle, app-specific recovery (D60): the oracle checks only universal
// legs (token, user, contract) — it does NOT know about the followers group.
// The followers-group heal is THIS app's concern, done client-side here (the
// social app is the one that knows what the followers group is).
//
// The load-bearing rule (definite NO vs. UNKNOWN): a verdict that is
// `inconclusive` (a check couldn't run — store unreadable) takes NO action.
// Only decisive negatives (contract missing, token dead, user gone) drive
// recovery. A deploy window must not look like "contract missing."

import { getV3Client } from './v3';
import type { AccessVerdict } from './v3';
import { ensureFollowers } from './groups';
import { getSocialAuth } from '../interfaces/auth';

const LOG = (...args: unknown[]) => console.log('[access]', ...args);
const LOG_ERR = (...args: unknown[]) => console.error('[access]', ...args);
// A transient "oracle unreachable" is a retry-later, not an error — warn, not
// error (console.error would trip the e2e console-error assertions).
const LOG_WARN = (...args: unknown[]) => console.warn('[access]', ...args);

// The services the social app needs — the recovery verifies these. The signal
// is platform-level (the node checks them); this list is the app's policy.
const SOCIAL_SERVICES = [
  'posts',
  'media',
  'public_media',
  'profile',
  'settings',
  'comments',
  'reactions',
  'contacts',
  'staging_posts',
];
const SOCIAL_OPERATIONS = ['readAll', 'create'];

// Cooldown: max one auto-recovery per action-class per window. After the
// window, the recovery stops acting and surfaces a manual "Log in again" signal
// (the loop-breaker — the system tries once, then hands the user the wheel).
const COOLDOWN_MS = 5 * 60 * 1000;
const lastRecovery: Record<string, number> = {};

export type AccessRecoveryResult =
  | { outcome: 'ok' }
  | { outcome: 'recovered'; actions: string[] }
  | { outcome: 'inconclusive' }
  | { outcome: 'needs_manual'; reason: string };

export interface VerifyAndRecoverOptions {
  /** When false (the on-mount call), a `reauth` (which opens the auth popup)
   *  is NOT auto-executed — it's deferred to an actual failure (the reactive
   *  path) or a user-initiated "Log in again". A popup on page load would be
   *  glitchy; a popup when an action fails is expected. The safe local actions
   *  (the followers-group heal, signout) still run on mount. */
  allowReauth?: boolean;
}

function inCooldown(action: string): boolean {
  const last = lastRecovery[action];
  return last !== undefined && Date.now() - last < COOLDOWN_MS;
}

function markRecovered(action: string): void {
  lastRecovery[action] = Date.now();
}

/** Test hook — clear the cooldown state. */
export function _resetRecoveryCooldown(): void {
  for (const k of Object.keys(lastRecovery)) delete lastRecovery[k];
}

/**
 * Verify access and execute the verdict's recovery actions.
 * Call at mount (when signed in, with allowReauth: false) and after a data
 * failure (allowReauth defaults true).
 */
export async function verifyAndRecover(
  options: VerifyAndRecoverOptions = {},
): Promise<AccessRecoveryResult> {
  const allowReauth = options.allowReauth ?? true;
  const w = getV3Client();
  const token = w.readToken();
  if (!token) {
    // Not signed in — the login screen owns this, not the recovery.
    LOG('verifyAndRecover — no token, skipping (login screen owns sign-in)');
    return { outcome: 'needs_manual', reason: 'not_signed_in' };
  }

  let verdict: AccessVerdict;
  try {
    verdict = await w.verifyAccess({ services: SOCIAL_SERVICES, operations: SOCIAL_OPERATIONS });
  } catch (e) {
    // The verify call itself failed (network / 5xx) — a transient infra
    // error, NOT a bad session. No action (definite-NO-vs-UNKNOWN).
    LOG_WARN('verifyAndRecover — verifyAccess unreachable (transient, no action):', e);
    return { outcome: 'inconclusive' };
  }

  LOG('verifyAndRecover — verdict:', JSON.stringify(verdict));

  if (verdict.status === 'inconclusive') {
    // A check couldn't run (store unreadable) — no action, retry later.
    LOG('verifyAndRecover — inconclusive, taking no action');
    return { outcome: 'inconclusive' };
  }

  const executed: string[] = [];

  // invalid — the token is dead (or the user is gone). Execute the decisive
  // actions (reauth / signout). No followers-group heal — a dead token can't
  // heal the group; the reauth re-derives a live session first.
  if (verdict.status === 'invalid') {
    for (const action of verdict.actions) {
      if (inCooldown(action)) {
        LOG(`verifyAndRecover — ${action} is in cooldown; deferring to manual`);
        return { outcome: 'needs_manual', reason: `cooldown:${action}` };
      }
      try {
        await executeAction(action);
        markRecovered(action);
        executed.push(action);
        LOG(`verifyAndRecover — executed ${action}`);
      } catch (e) {
        LOG_ERR(`verifyAndRecover — ${action} failed:`, e);
        return { outcome: 'needs_manual', reason: `action_failed:${action}` };
      }
    }
    return executed.length ? { outcome: 'recovered', actions: executed } : { outcome: 'ok' };
  }

  // ok or degraded — the token is valid. First the app's own followers-group
  // heal (D60: the oracle is generic, so the social app ensures its OWN group
  // here — idempotent, a no-op when healthy). It runs BEFORE the reauth
  // deferral check so it still runs on mount (the token is valid there).
  if (!inCooldown('heal_followers_group')) {
    try {
      LOG('verifyAndRecover — ensuring the followers group for', token.username);
      await ensureFollowers(token.username, token.provider);
      markRecovered('heal_followers_group');
      executed.push('heal_followers_group');
    } catch (e) {
      LOG_ERR('verifyAndRecover — followers-group heal failed:', e);
      return { outcome: 'needs_manual', reason: 'action_failed:heal_followers_group' };
    }
  }

  // Then the oracle's actions (reauth for a degraded contract; none for ok).
  for (const action of verdict.actions) {
    if (action === 'reauth' && !allowReauth) {
      // On mount, don't open the popup — surface a soft "log in again" signal
      // and let an actual failure (or the user) trigger the reauth.
      LOG('verifyAndRecover — reauth deferred (mount); surfacing a manual signal');
      return { outcome: 'needs_manual', reason: 'reauth_deferred' };
    }
    if (inCooldown(action)) {
      LOG(`verifyAndRecover — ${action} is in cooldown; deferring to manual`);
      return { outcome: 'needs_manual', reason: `cooldown:${action}` };
    }
    try {
      await executeAction(action);
      markRecovered(action);
      executed.push(action);
      LOG(`verifyAndRecover — executed ${action}`);
    } catch (e) {
      LOG_ERR(`verifyAndRecover — ${action} failed:`, e);
      return { outcome: 'needs_manual', reason: `action_failed:${action}` };
    }
  }

  return executed.length ? { outcome: 'recovered', actions: executed } : { outcome: 'ok' };
}

async function executeAction(action: string): Promise<void> {
  const auth = getSocialAuth();
  switch (action) {
    case 'reauth': {
      // Re-derive through the rooted authenticator (fresh token + contract).
      // REPLACE-ON-ARRIVAL: the handed-back token overwrites the stale cookie
      // (authListen does this naturally). We do NOT scrub first — a blocked
      // popup or a transient failure must not strand the user signed-out.
      LOG('reauth — re-deriving the session via the authenticator');
      auth.login();
      return;
    }
    case 'signout': {
      // Terminal — a deleted account can't be re-authed. Clear the session and
      // let the app show the login screen.
      LOG('signout — terminal (user not found), clearing the session');
      auth.signOut();
      window.dispatchEvent(
        new CustomEvent('session:signed-out', { detail: { reason: 'user_not_found' } }),
      );
      return;
    }
    default:
      LOG(`executeAction — unknown action "${action}", ignoring`);
  }
}

File: api/app/v3/endpoints/access.py
"""POST /v3/session/verify — the confirmatory session-health oracle.

The app's SessionGuard calls this at mount and after a failure. Instead of
guessing from status codes (a 401 here means "bad token" OR "no permission"
OR "user not found" — the info is destroyed at the boundary), the server runs
the ACTUAL checks it would run on a real request and reports each as a stable
code, plus the ordered recovery `actions` the client should execute.

The load-bearing rule (definite NO vs. UNKNOWN): every store-backed field
separates a DECISIVE answer (the check ran clean and found nothing) from
`unknown` (the check couldn't run — the store was unreadable). Only decisive
negatives drive actions. A deploy window that takes the contract store down
yields `unknown`, NOT `missing` — so the guard stays idle instead of churning
every user into a re-auth loop.
"""

import logging
from datetime import datetime

import jwt
from fastapi import APIRouter, Request

import app.settings as settings
from app.services.auth import decode_token
from app.v3.models import VerifySession
from app.v3.services import clickhouse as ch

logger = logging.getLogger(__name__)

router = APIRouter(tags=["access"])


def _check_token(token: str | None) -> tuple[str, str | None, str | None]:
    """Decode + classify the token. Returns (state, username, provider).

    state: valid | expired | invalid | missing. username/provider are only
    meaningful when state == "valid".
    """
    if not token:
        return "missing", None, None
    try:
        decoded = decode_token(token, private_key=True)
    except jwt.ExpiredSignatureError:
        return "expired", None, None
    except jwt.InvalidTokenError:
        # Bad signature, malformed, wrong algorithm — all "invalid".
        return "invalid", None, None
    if not decoded.username or decoded.username == "anon":
        # Well-formed but not a user session.
        return "invalid", None, None
    # User tokens carry a custom `expires` claim (ISO string), NOT the standard
    # `exp` — PyJWT doesn't check it, so check it here. This mirrors the SDK's
    # isTokenExpired (the client-side check); the server otherwise would report
    # a lapsed token as "valid".
    if decoded.expires:
        try:
            if datetime.utcnow() >= datetime.fromisoformat(decoded.expires):
                return "expired", None, None
        except ValueError:
            pass  # unparseable expires — signature still valid, not expired
    provider = decoded.provider or settings.PROVIDER
    return "valid", decoded.username, provider


@router.post("/verify")
def verify_access(request: Request, data: VerifySession):
    """Return a typed access verdict + ordered recovery actions.

    Generic by design (D60): checks only universal legs — token, user,
    contract. It does NOT know about any app's groups (no followers-group
    check); app-specific recovery (e.g. the social app healing its followers
    group) is the app's job, client-side.
    """
    token_state, username, provider = _check_token(data.token)

    # user: exists | not_found | unknown (only checkable with a valid token)
    user_state = "unknown"
    # contract: granted | partial | missing | unknown | not_checked
    # ("not_checked" = the app declared no services — a health probe; it must
    # not taint the verdict, so it's excluded from the degraded/unknown logic)
    contract_state = "not_checked"
    contract_checked = False
    missing_services: list[str] = []

    if token_state == "valid":
        # --- user lookup (decisive unless the store is unreadable) ---
        try:
            user_state = "exists" if ch.get_user(username) else "not_found"
        except Exception as e:  # store unreadable → unknown, NOT not_found
            logger.warning("[session] verify: user lookup unreadable: %s", e)
            user_state = "unknown"

        # --- app-contract check (needs the request Origin) ---
        if data.services:
            contract_checked = True
            origin = request.headers.get("origin", "")
            try:
                perms = ch.get_app_permissions(username, origin)
                missing_services = [
                    svc for svc in data.services if not all(op in perms.get(svc, []) for op in data.operations)
                ]
                if not missing_services:
                    contract_state = "granted"
                elif len(missing_services) < len(data.services):
                    contract_state = "partial"
                else:
                    contract_state = "missing"
            except Exception as e:  # store unreadable → unknown, NOT missing
                logger.warning("[session] verify: contract check unreadable: %s", e)
                contract_state = "unknown"
                missing_services = []

    # --- status: invalid > degraded > inconclusive > ok ---
    # Any non-valid token (expired / invalid / missing) is a dead session.
    token_dead = token_state in ("expired", "invalid", "missing")
    contract_bad = contract_checked and contract_state in ("missing", "partial")
    contract_unknown = contract_checked and contract_state == "unknown"
    if user_state == "not_found" or token_dead:
        status = "invalid"
    elif contract_bad:
        status = "degraded"
    elif user_state == "unknown" or contract_unknown:
        status = "inconclusive"
    else:
        status = "ok"

    # --- actions: decisive problems only, lightest/needed-first ---
    if user_state == "not_found":
        # Re-auth can't resurrect a deleted account — terminal sign-out.
        actions: list[str] = ["signout"]
    else:
        actions = []
        # Re-derive through the rooted authenticator (fresh token + contract).
        if token_dead or contract_bad:
            actions.append("reauth")

    return {
        "status": status,
        "token": token_state,
        "user": user_state,
        "contract": {"state": contract_state, "missing_services": missing_services},
        "actions": actions,
        "username": username,
        "provider": provider,
    }
